blockchain/stake/treasury_test.go

```diff
@@ -0,0 +1,1740 @@
+				0x0c, // OP_DATA_12
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00,
+			},
+		},
+	},
+	LockTime: 0,
+	Expiry:   0,
+}
+
+// treasurybaseInvalidOpcode1 has an invalid out script opcode.
+var treasurybaseInvalidOpcode1 = &wire.MsgTx{
+	SerType: wire.TxSerializeFull,
+	Version: 3,
+	TxIn: []*wire.TxIn{
+		{},
+	},
+	TxOut: []*wire.TxOut{
+		{
+			PkScript: []byte{
+				0xc1, // OP_TADD
+			},
+		},
+		{
+			PkScript: []byte{
+				0xc1, // OP_TADD instead of OP_RETURN
+				0x0c, // OP_DATA_32
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00,
+			},
+		},
+	},
+	LockTime: 0,
+	Expiry:   0,
+}
+
+// treasurybaseInvalidOpcode1Len has an invalid out script opcode length.
+var treasurybaseInvalidOpcode1Len = &wire.MsgTx{
+	SerType: wire.TxSerializeFull,
+	Version: 3,
+	TxIn: []*wire.TxIn{
+		{},
+	},
+	TxOut: []*wire.TxOut{
+		{
+			PkScript: []byte{
+				0xc1, // OP_TADD
+			},
+		},
+		{
+			PkScript: nil,
+		},
+	},
+	LockTime: 0,
+	Expiry:   0,
+}
+
+// treasurybaseInvalidOpcodeDataPush has an invalid out script data push in
+// script 1 opcode 1.
+var treasurybaseInvalidOpcodeDataPush = &wire.MsgTx{
+	SerType: wire.TxSerializeFull,
+	Version: 3,
+	TxIn: []*wire.TxIn{
+		{},
+	},
+	TxOut: []*wire.TxOut{
+		{
+			PkScript: []byte{
+				0xc1, // OP_TADD
+			},
+		},
+		{
+			PkScript: []byte{
+				0x6a, // OP_RETURN
+				0x05, // OP_DATA_5 instead of OP_DATA_4
+				0x00, 0x00, 0x00, 0x00, 0x00,
+			},
+		},
+	},
+	LockTime: 0,
+	Expiry:   0,
+}
+
+// treasurybaseInvalid has invalid in script constants.
+var treasurybaseInvalid = &wire.MsgTx{
+	SerType: wire.TxSerializeFull,
+	Version: 3,
+	TxIn: []*wire.TxIn{
+		{
+			PreviousOutPoint: wire.OutPoint{
+				Index: math.MaxUint32 - 1,
+			},
+		},
+	},
+	TxOut: []*wire.TxOut{
+		{
+			PkScript: []byte{
+				0xc1, // OP_TADD
+			},
+		},
+		{
+			PkScript: []byte{
+				0x6a, // OP_RETURN
+				0x0c, // OP_DATA_12
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00,
+			},
+		},
+	},
+	LockTime: 0,
+	Expiry:   0,
+}
+
+// treasurybaseInvalid2 has invalid in script constants.
+var treasurybaseInvalid2 = &wire.MsgTx{
+	SerType: wire.TxSerializeFull,
+	Version: 3,
+	TxIn: []*wire.TxIn{
+		{
+			PreviousOutPoint: wire.OutPoint{
+				Index: math.MaxUint32,
+				Hash:  chainhash.Hash{'m', 'o', 'o'},
+			},
+		},
+	},
+	TxOut: []*wire.TxOut{
+		{
+			PkScript: []byte{
+				0xc1, // OP_TADD
+			},
+		},
+		{
+			PkScript: []byte{
+				0x6a, // OP_RETURN
+				0x0c, // OP_DATA_12
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00,
+			},
+		},
+	},
+	LockTime: 0,
+	Expiry:   0,
+}
+
+// treasurybaseInvalidTxVersion has an invalid transaction version.
+var treasurybaseInvalidTxVersion = &wire.MsgTx{
+	SerType: wire.TxSerializeFull,
+	Version: 1, // Invalid
+	TxIn: []*wire.TxIn{
+		{
+			PreviousOutPoint: wire.OutPoint{
+				Index: math.MaxUint32,
+				Hash:  chainhash.Hash{'m', 'o', 'o'},
```

Comment:
Still also invalid because of the Hash.

Reply:
TxVersion is being tested which comes earlier. And that is totally a valid hash. It may not match the transaction but the hash is valid.